File: 94.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        stack = []
        tem = root
        while 1:
            self.go_along_left(stack, tem)
            if not stack:
                break
            tem = stack.pop()
            res.append(tem.val)
            tem = tem.right

        return res

    def go_along_left(self, stack, root):
        while root:
            stack.append(root)
            root = root.left

    def inorderTraversal2(self, root):
        ans = []
        while root:
            if root.left:
                pre = root.left
                while pre.right and pre.right is not root:
                    pre = pre.right
                if pre.right is root:
                    ans.append(root.val)
                    pre.right = None
                    root = root.right
                else:
                    pre.right = root
                    root = root.left
            else:
                ans.append(root.val)
                root = root.right
        return ans    

File: 94.go
package leetcode

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func inorderTraversal2(root *TreeNode) (ans []int) {
	if root == nil {
		return
	}
	stack := []interface{}{root}
	for l := len(stack); l > 0; l = len(stack) {
		node := stack[l-1]
		stack = stack[:l-1]
		if val, ok := node.(int); ok {
			ans = append(ans, val)
		} else {
			treeNode, _ := node.(*TreeNode)
			if treeNode.Right != nil {
				stack = append(stack, treeNode.Right)
			}
			stack = append(stack, treeNode.Val)
			if treeNode.Left != nil {
				stack = append(stack, treeNode.Left)
			}
		}
	}
	return ans
}

func inorderTraversal(root *TreeNode) (ans []int) {
	if root == nil {
		return
	}
	stack := []*TreeNode{}

	goLeft := func(root *TreeNode) {
		// 有了这个东西出栈的节点都可以直接把值加到ans中
		for root != nil {
			stack = append(stack, root)
			root = root.Left
		}
	}

	goLeft(root)

	for l := len(stack); l > 0; l = len(stack) {
		node := stack[l-1]
		stack = stack[:l-1]
		ans = append(ans, node.Val)
		goLeft(node.Right)
	}

	return ans
}
